# Projet 3 (LU3IN005) : Chaines de Markov et épidémiologie :
# Propagation d’une épidémie dans une population

### <u>Groupe 6 :</u>
<pr>Samy (Adem) YAZID<br>Numéro étudiant : $3811684$<br><br>Margaux MARSELOO<br>Numéro étudiant : $28630611$</pr> 

### <u>**Consignes :**</u>
<pr>

L’objectif de ce projet est de manipuler des chaînes de Markov pour étudier la propagation d’une épidémie dans une population.<br>
Notre rendu est un notebook contenant les codes commentés et les résultats interprétés.<br>
Les packages <em>numpy</em>, <em>random</em> et <em>matplotlib</em> sont utilisés.<br>
Nous allons étudier des populations constituées de $3$ types d’individus.<br>
Chaque individu est dans un des $3$ états suivant : sain $S$, infecté $I$ ou guéri $R$.
</pr>

In [113]:
import numpy as np
import math
import matplotlib.pyplot as plt
import random as rd

def rec_extract_values(iterable, dico):
    """
    Paramètres:
    iterable : any
        Iterable (ou non) dont on veut extraire les valeurs dans un dictionnaire
    dico : dict
        Dictionnaire qui stockera les valeurs extraites
        
    Retourne:
    None

    Description:
    Fonction permettant d'extraire toutes les valeurs de iterable en les stockant
    dans dico, et qui compte toutes les fois où chaque valeur est apparue.
    (Dans cette fonction les strings sont considérés comme non-iterable, donc
    seront compté comme une valeur quelconque)
    
    Exemple:
    dico : {}
    rec_extract_values([1, 3, [3, 7]])
    dico : {1:1, 3:2, 7:1}
    """
    if type(iterable) in [type([]), type(np.matrix([])), type(np.array([]))]:
        for l in iterable:
            rec_extract_values(l, dico)
        return dico
    else:
        if iterable not in dico:
            dico[iterable] = 1
        else:
            dico[iterable] += 1
            
def rec_sum_values(iterable):
    """
    Paramètres:
    iterable : any iterable type of (int or float)
        Iterable dont on veut sommer les valeurs
        
    Retourne:
    somme : int or float
        Somme des valeurs de iterable
    
    Description:
    Fonction permettant de retourner la somme de toutes les valeurs de iterable.
    
    Exemple:
    rec_sum_values([1, 3, [3, 7.5]]) : 14.5
    """
    if type(iterable) in [type([]), type(np.matrix([])), type(np.array([]))]:
        somme = 0
        for l in iterable:
            somme += rec_sum_values(l)
        return somme
    else:
        return iterable
    
def dict_zero(dico):
    """
    Paramètres:
    dico : dict
        Dictionnaire dont on veut mettre toutes ses valeurs à 0
        
    Retourne:
    None
    
    Description:
    Fonction permettant pour un dictionnaire de conserver ses clés tout en fixant leur valeur à 0.
    
    Exemple:
    dico : {1:1, 3:2, 7:1}
    dict_zero(dico)
    dico : {1:0, 3:0, 7:0}
    """
    for key in dico:
        dico[key] = 0

## I) <u>Apprentissage des paramètres d’un modèle à partir de données</u>

<pr>
    
Dans notre modèle, nous allons considérer qu’à chaque temps :
- Chaque individu sain peut rester sain ou devenir infecté
- Chaque individu infecté peut rester infecté ou devenir guéri
- Chaque individu guéri reste guéri
et que la probabilité de passer d’un état à l’autre ne dépend que de l’état précédent.

Nous disposons d’une séquence d’observations et nous souhaitons apprendre les paramètres de la
chaîne de Markov permettant de modéliser le processus sous-jacent qui a généré la séquence.<br>
Nous avons suivi un individu pendant $10$ jours, afin de déterminer a chaque temps dans quel état se trouvait l’individu.<br>
Nous avons obtenu la séquence d’observation suivante : $S, S, S, I, I ,I ,I , I, I, R$.
</pr>


<pr><u>**Question 1:**</u> À partir de cette séquence d’observation, estimez de les probabilités de transition entre les états et dresser la matrice de probabilité de transitions.</pr>


Matrice de transition du modèle :
$
\begin{pmatrix}
  &   S &   I &   R \\
S & 2/3 & 1/3 &   0 \\
I &   0 & 5/6 & 1/6 \\
R &   0 &   0 &   1 \\
\end{pmatrix}
$

In [112]:
matrice_transition = np.array([[2/3, 1/3,   0], 
                               [  0, 5/6, 1/6], 
                               [  0,   0,   1]])

<pr>
Nous avons ensuite suivi une population de $5000$ individus, pendant $200$ jours.<br>
Pour lire les données, nous avons utiliser <code>np.loadtxt("data_exo_2022.txt")</code>.<br>
Les individus sains sont notés $0$, les infectés $1$ et les guéris $2$.

<u>**Question 2:**</u><br>
    
1) Lire des données
    
</pr>

In [114]:
data = np.loadtxt("data_exo_2022.txt")

<pr>
    
2) Estimez les probabilités de transition entre les états et dressez la matrice de probabilité de transitions
(indice pour vérifier vos résultats : la première ligne de la matrice est $[0.9308, 0.0691, 0.]$).

</pr>

In [115]:
def analyse_seq(sequence, nb_etats=-1):
    sequence = sequence.astype(int)
    
    if nb_etats == -1:
        card = {elem : 0 for elem in set(sequence)}
        nb_etats = len(card)
        mat_transition_etats = np.zeros((nb_etats, nb_etats))
    else:
        mat_transition_etats = np.zeros((nb_etats, nb_etats))
    
    #on détermine la fréquence
    for etat in range(1, len(sequence)):
        mat_transition_etats[sequence[etat-1]][sequence[etat]] += 1
    mat_transition_etats[sequence[etat]][sequence[etat]] += 1
        
    return mat_transition_etats  

def matrice_proba_transition_liste(liste_sequence):
    card = {}
    rec_extract_values(liste_sequence, card)
    matrice_finale = np.zeros((len(card), len(card)))
    
    for seq in liste_sequence:
        matrice_finale += analyse_seq(seq, len(card))
    print("\nMatrice des fréquences :")
    print(matrice_finale)
    print("\n")
    print("card[0] = ",card[0])
    for i in range(len(matrice_finale)):
        for j in range(len(matrice_finale[0])):
            matrice_finale[i][j] /= card[i]
            
    return matrice_finale

print("Matrice de transition de l'exemple:\n")
print(matrice_proba_transition_liste([np.array([0., 0., 0., 1., 1., 1., 1., 1., 1., 2.])]))
print("\nMatrice de transition des 5000 individus:\n")
print(matrice_proba_transition_liste(data))

Matrice de transition de l'exemple:


Matrice des fréquences :
[[2. 1. 0.]
 [0. 5. 1.]
 [0. 0. 1.]]


card[0] =  3
[[0.66666667 0.33333333 0.        ]
 [0.         0.83333333 0.16666667]
 [0.         0.         1.        ]]

Matrice de transition des 5000 individus:


Matrice des fréquences :
[[274606.  20308.      0.]
 [     0. 274803.  18953.]
 [ 16630.      0. 394700.]]


card[0] =  294914
[[0.93113925 0.06886075 0.        ]
 [0.         0.93548047 0.06451953]
 [0.04042983 0.         0.95957017]]


Matrice de fréquence finale:
$
\begin{pmatrix}
  &   S &   I &   R \\
S & 274506 & 20408 &   0 \\
I &   0 & ? & ? \\
R &   0 &   0 &  0 \\
\end{pmatrix}
$

## II) <u>Description du premier modèle</u>
<pr>La probabilité pour un individu d’être dans un de ces $3$ états au temps $t$, ne dépend que l’état dans
lequel il est au temps $t − 1$.<br>
Un individu dans l’état sain a une probabilité de $0.92$ de rester sain et une probabilité de $0.08$ de
devenir infecté. Si l’individu est infecté, il peut le rester avec une probabilité de $0.93$ et être guéri avec
une probabilité de $0.07$. S’il est dans l’état guéri, il reste dans cet état avec une probabilité de $1$.

<u>**Question 1:**</u> À partir du graphe de transition, créez la matrice de transition $A$, la matrice contenant les probabilités de transition entre les différents états. Créez une fonction permettant de vérifier qu’une matrice est stochastique.
</pr>

Matrice de transition du premier modèle :<br>
$A = 
\begin{pmatrix}
  &    S &    I &    R \\
S & 0.92 & 0.08 &    0 \\
I &    0 & 0.93 & 0.07 \\
R &    0 &    0 &    1 \\
\end{pmatrix}
$

In [116]:
matrice_transition_modele1 = np.array([[0.92, 0.08,    0], 
                                       [   0, 0.93, 0.07], 
                                       [   0,    0,    1]])

def est_stochastique(matrice):
    epsilon = 0.00001
    for ligne in matrice:
        if abs(rec_sum_values(ligne) - 1) >= epsilon:
            return False
    return True

print("La matrice de transition de la partie I) est-elle stochastique ? :", est_stochastique(matrice_transition))

print("La matrice de transition de la partie II) est-elle stochastique ? :", est_stochastique(matrice_transition))

La matrice de transition de la partie I) est-elle stochastique ? : True
La matrice de transition de la partie II) est-elle stochastique ? : True


<pr>
Au temps $t = 0$, un individu a une probabilité de $0.9$ d’être sain et $0.1$ d’être infecté.

<u>**Question 2:**</u> Créez $π_0$ la distribution de probabilité initiale.
</pr>

$
π_0 = 
\begin{pmatrix}
   S &    I &    R \\
0.90 & 0.10 &    0 \\
\end{pmatrix}
$

In [117]:
pi_0 = np.array([0.90,
                 0.10,
                 0.00])

<pr>
Nous allons étudier l’évolution du nombre d’individu sains, infectés et guéris au cours du temps.
Dans un premier temps nous étudierons la distribution théorique, puis la distribution observée sur des simulations.
</pr>

### <u>Distribution $π_t$<u>

<pr><u>**Question 1:**</u> En utilisant $π_0$ et $A$, donnez la probabilité pour un individu d’être sain, infecté ou guéri au temps $t = 1$ (faire d’abord le calcul à la main).
</pr>


$
π_1 = π_0 * A =
\begin{pmatrix}
   S &    I &    R \\
0.90*0.92 + 0*0 & 0.90*0.08 + 0.10*0.93 &    0.10*0.07 + 0*1 \\
\end{pmatrix}
=
\begin{pmatrix}
   S &    I &    R \\
0.828 & 0.165 & 0.007\\
\end{pmatrix}
$

In [118]:
pi_1 = np.dot(pi_0, matrice_transition_modele1) 

print("pi_1 =", pi_1)

pi_1 = [0.828 0.165 0.007]


<pr><u>**Question 2:**</u> Donnez la probabilité pour un individu d’être sain, infecté ou guéri au temps $t = 2$ (faire d’abord le calcul à la main).
</pr>

$
π_2 = π_1 * A =
\begin{pmatrix}
   S &    I &    R \\
0.828*0.92 + 0.007*0   &   0.828*0.08 + 0.165*0.93   &   0.165*0.07 + 0.007*1 \\
\end{pmatrix}
=
\begin{pmatrix}
   S &    I &    R \\
0.76176 & 0.21969 & 0.01855\\
\end{pmatrix}
$

In [119]:
pi_2 = np.dot(pi_1, matrice_transition_modele1) 

print("pi_2 =", pi_2)


pi_2 = [0.76176 0.21969 0.01855]


<pr><u>**Question 3:**</u> De même pour chaque temps $t$ entre $1$ et $200$, calculez la distribution théorique des effectifs dans chaque état (<em>Rappel:</em> $π_{t+1} = π_t . A$ ).
</pr>

In [120]:
def liste_pi(pi_0, A, n):
    l_pi = [pi_0]
    for i in range(1, n+1):
        l_pi.append(np.dot(l_pi[i-1], matrice_transition_modele1))
    return l_pi

liste_pi_i = np.array(liste_pi(pi_0, matrice_transition_modele1, 200))
print("Liste des pi_i pour i entre 0 et 200 :")
print(liste_pi_i)

Liste des pi_i pour i entre 0 et 200 :
[[9.00000000e-01 1.00000000e-01 0.00000000e+00]
 [8.28000000e-01 1.65000000e-01 7.00000000e-03]
 [7.61760000e-01 2.19690000e-01 1.85500000e-02]
 [7.00819200e-01 2.65252500e-01 3.39283000e-02]
 [6.44753664e-01 3.02750361e-01 5.24959750e-02]
 [5.93173371e-01 3.33138129e-01 7.36885003e-02]
 [5.45719501e-01 3.57272330e-01 9.70081693e-02]
 [5.02061941e-01 3.75920827e-01 1.22017232e-01]
 [4.61896986e-01 3.89771324e-01 1.48331690e-01]
 [4.24945227e-01 3.99439090e-01 1.75615683e-01]
 [3.90949609e-01 4.05473972e-01 2.03576419e-01]
 [3.59673640e-01 4.08366763e-01 2.31959597e-01]
 [3.30899749e-01 4.08554980e-01 2.60545271e-01]
 [3.04427769e-01 4.06428112e-01 2.89144119e-01]
 [2.80073547e-01 4.02332365e-01 3.17594087e-01]
 [2.57667664e-01 3.96574984e-01 3.45757353e-01]
 [2.37054251e-01 3.89428148e-01 3.73517602e-01]
 [2.18089911e-01 3.81132518e-01 4.00777572e-01]
 [2.00642718e-01 3.71900434e-01 4.27456848e-01]
 [1.84591300e-01 3.61918821e-01 4.53489879e-01]
 

<pr><u>**Question 4:**</u> Représentez graphiquement la probabilité d’être dans chaque état en fonction du temps. (+ décrivez un peu ce que vous observez).
</pr>

In [121]:
def affiche_graphe_multi(liste_pi_i, xlabel="Temps t", ylabel="Probabilités"):
    #on crée les ordonnées
    liste_sains = []
    liste_infectes = []
    liste_gueris = []
    for pi in liste_pi_i:
        liste_sains.append(pi[0])
        liste_infectes.append(pi[1])
        liste_gueris.append(pi[2])
    
    #on crée les abscisses
    abscisses = [i for i in range(len(liste_pi_i))]
    
    #on plot le graphe comme une courbe
    plt.plot(abscisses, liste_sains)
    plt.plot(abscisses, liste_infectes)
    plt.plot(abscisses, liste_gueris)
    
    #on légende le graphe
    plt.legend(["Personnes saines", "Personnes infectées", "Personnes guéries"], loc='center right')
    
    #on nomme l'axe des abscisses
    plt.xlabel(xlabel)
    #on nomme l'axe des ordonnées
    plt.ylabel(ylabel)
    
    #on paramètre l'échelle des abscisses suivant l'option x_scale
    plt.xscale("linear")
    #on paramètre l'échelle des ordonnées comme linéaire
    plt.yscale("linear")
    
    #on donne un titre au graphique
    plt.title("Graphique des proportions la population en fonction du temps")
       
    #on affiche le graphe
    plt.show()
    
#affiche_graphe_multi(liste_pi_i)

<img src="Graph_Partie2.png"  align="bottom">

<pr><u>**Interprétation du graphique:**</u>
    Nous remarquons qu'il n'y qu'une seule vague d'infection. Les états sains et infectés sont transients tant dit que l'état guéri est récurrent. L'état guéri est en effet un état absorbant. Nous voyons bien ici que ce modèle converge rapidement vers 
$π ^* = 
\begin{pmatrix}
   S &    I &     R \\
 0.0 &  0.0 & 1.0 \\
\end{pmatrix}
$
.<br>
</pr>

### <u>Tirage aléatoire des états</u>
<pr>Nous allons générer une séquence de taille $T$ en utilisant cette chaîne de Markov. Pour générer une séquence aléatoire, nous choisissons un état initial au hasard (en utilisant $π_0$) ; puis nous choisissons les états suivants en suivant les probabilités de transition <em>(= la matrice de transition $A$).</em>
Nous prenons $T=150$.
</pr>

In [122]:
def genere_sequence(pi_0, A, n):
    p = rd.random()
    sequence = []
    
    #on choisit l'état initial au hasard
    for i in range(len(pi_0)):
        if p < sum(pi_0[:i+1]):
            sequence.append(i)
            break

    #on choisit les états suivants
    for i in range(1, n):
        p = rd.random()
        for j in range(len(A)):
            if p < sum(A[sequence[-1]][:j+1]):
                sequence.append(j)
                break
    return np.array(sequence)

print("L'une des séquences aléatoires de taille 150 suivant cette chaîne de Markov est :\n")
print(genere_sequence(pi_0, matrice_transition_modele1, 150))

L'une des séquences aléatoires de taille 150 suivant cette chaîne de Markov est :

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


### <u>Modélisation d’une population</u>
<pr>Nous avons généré une séquence d’état pour un individu. Maintenant nous allons générer un ensemble de séquences pour une population de 200 individus.

<u>**Question 1:**</u> À chaque temps $t$, nous comptons le nombre d’individus sains, infectés et guéris dans la population et affichons l’évolution du nombre d’individus dans les trois états en fonction du temps.
</pr>

In [123]:
def genere_liste_sequence(pi_0, A, nb_individu, temps):
    liste_sequence = []
    for i in range(nb_individu):
        liste_sequence.append(genere_sequence(pi_0, A, temps))
    return np.array(liste_sequence)

def comptage_liste_sequence(liste_sequence, nb_etats=3):
    liste_comptage = []
    for jour in range(len(liste_sequence[0])):
        comptage = [0 for i in range(nb_etats)]
        for seq in liste_sequence:
            comptage[seq[jour]] += 1
        liste_comptage.append(comptage)
    return np.array(liste_comptage)

liste_sequence = genere_liste_sequence(pi_0, matrice_transition_modele1, 200, 150)
liste_comptage = comptage_liste_sequence(liste_sequence)

print("La liste de 200 séquences est :\n")
print(liste_sequence)

#EST CE QUE çA SERT D'AFFICHER çA ??
#Je sais pas mais c'est pour montrer que la fonction marche x)
print("\n\nLa liste de 200 séquences compte cet effectif au cours du temps:\n")
print(liste_comptage)

La liste de 200 séquences est :

[[0 0 0 ... 2 2 2]
 [0 0 1 ... 2 2 2]
 [0 1 1 ... 2 2 2]
 ...
 [0 0 0 ... 2 2 2]
 [0 0 0 ... 2 2 2]
 [0 0 0 ... 2 2 2]]


La liste de 200 séquences compte cet effectif au cours du temps:

[[182  18   0]
 [165  35   0]
 [154  43   3]
 [139  56   5]
 [127  64   9]
 [115  71  14]
 [101  81  18]
 [ 93  84  23]
 [ 85  86  29]
 [ 80  83  37]
 [ 75  83  42]
 [ 72  82  46]
 [ 70  82  48]
 [ 60  84  56]
 [ 53  87  60]
 [ 51  82  67]
 [ 49  80  71]
 [ 47  78  75]
 [ 44  73  83]
 [ 41  67  92]
 [ 39  59 102]
 [ 36  59 105]
 [ 33  55 112]
 [ 31  53 116]
 [ 31  53 116]
 [ 28  49 123]
 [ 26  46 128]
 [ 24  45 131]
 [ 21  44 135]
 [ 19  45 136]
 [ 19  41 140]
 [ 16  41 143]
 [ 15  38 147]
 [ 14  36 150]
 [ 13  37 150]
 [ 10  39 151]
 [ 10  38 152]
 [  9  38 153]
 [  8  37 155]
 [  8  33 159]
 [  7  30 163]
 [  6  28 166]
 [  6  26 168]
 [  6  22 172]
 [  5  22 173]
 [  5  20 175]
 [  4  19 177]
 [  3  18 179]
 [  3  16 181]
 [  3  15 182]
 [  3  15 182]
 [  3  15 182]

In [124]:
def affiche_graphe_nb_individu(liste_comptage):
    #on crée les ordonnées
    liste_sains = []
    liste_infectes = []
    liste_gueris = []
    
    for occurence in liste_comptage:
        liste_sains.append(occurence[0])
        liste_infectes.append(occurence[1])
        liste_gueris.append(occurence[2])
    
    #on crée les abscisses
    abscisses = [i for i in range(len(liste_comptage))]
    
    #on plot le graphe comme une courbe
    plt.plot(abscisses, liste_sains)
    plt.plot(abscisses, liste_infectes)
    plt.plot(abscisses, liste_gueris)
    
    #on légende le graphe
    plt.legend(["Personnes saines", "Personnes infectées", "Personnes guéries"], loc='center right')
    
    #on nomme l'axe des abscisses
    plt.xlabel("Temps t")
    #on nomme l'axe des ordonnées
    plt.ylabel("Nombre d'individu")
    
    #on paramètre l'échelle des abscisses suivant l'option x_scale
    plt.xscale("linear")
    #on paramètre l'échelle des ordonnées comme linéaire
    plt.yscale("linear")
    
    #on donne un titre au graphique
    plt.title("Graphique des états des individus en fonction du temps")
       
    #on affiche le graphe
    plt.show()
    
#affiche_graphe_nb_individu(liste_comptage)

<img src="Graph_Partie2_2.png"  align="bottom">

<pr><u>**Question 2:**</u> Affichez le pourcentage d’individus sains infectés et guéris en fonction du temps.
</pr>

In [126]:
def pourcentage_liste_sequence(liste_sequence):
    liste_pourcentage = [[((cpt/sum(l_cpt))*100) for cpt in l_cpt] for l_cpt in comptage_liste_sequence(liste_sequence)]
    return np.array(liste_pourcentage)

liste_pourcentage = pourcentage_liste_sequence(liste_sequence)
print("\n\nLa liste de 200 séquences compte ces pourcentage d'individus sains/infectés/guéris au cours du temps:")
print(liste_pourcentage)

##BIZARRE ! AU DEBUT ON A PAS 100% de SAINS DANS LE COMPTE ET ICI ON A ...
##Lorsque j'ai exécuté le code y a vraiment pas 100% de sains que ce soit ici ou dans le graphe
##(J'obtiens [ 91.    9.    0. ] en première ligne et j'ai pareil sur le graphe d'avant)



La liste de 200 séquences compte ces pourcentage d'individus sains/infectés/guéris au cours du temps:
[[ 91.    9.    0. ]
 [ 82.5  17.5   0. ]
 [ 77.   21.5   1.5]
 [ 69.5  28.    2.5]
 [ 63.5  32.    4.5]
 [ 57.5  35.5   7. ]
 [ 50.5  40.5   9. ]
 [ 46.5  42.   11.5]
 [ 42.5  43.   14.5]
 [ 40.   41.5  18.5]
 [ 37.5  41.5  21. ]
 [ 36.   41.   23. ]
 [ 35.   41.   24. ]
 [ 30.   42.   28. ]
 [ 26.5  43.5  30. ]
 [ 25.5  41.   33.5]
 [ 24.5  40.   35.5]
 [ 23.5  39.   37.5]
 [ 22.   36.5  41.5]
 [ 20.5  33.5  46. ]
 [ 19.5  29.5  51. ]
 [ 18.   29.5  52.5]
 [ 16.5  27.5  56. ]
 [ 15.5  26.5  58. ]
 [ 15.5  26.5  58. ]
 [ 14.   24.5  61.5]
 [ 13.   23.   64. ]
 [ 12.   22.5  65.5]
 [ 10.5  22.   67.5]
 [  9.5  22.5  68. ]
 [  9.5  20.5  70. ]
 [  8.   20.5  71.5]
 [  7.5  19.   73.5]
 [  7.   18.   75. ]
 [  6.5  18.5  75. ]
 [  5.   19.5  75.5]
 [  5.   19.   76. ]
 [  4.5  19.   76.5]
 [  4.   18.5  77.5]
 [  4.   16.5  79.5]
 [  3.5  15.   81.5]
 [  3.   14.   83. ]
 [  3.   13.  

In [125]:
#affiche_graphe_multi(liste_pourcentage)

<img src="Graph_Partie2_3.png"  align="bottom">

<pr><u>**Question 3:**</u> Quand $t$ est grand, quelle est la proportion d’individus sains, infectés et guéris ?
</pr>

<pr>
Lorsque $t$ est grand, la répartition de la population est la suivante :  π* = 
$ 
\begin{pmatrix}
   S &    I &     R \\
 0\% &  0\% & 100\% \\
\end{pmatrix}
$
.<br>
</pr>

<pr><u>**Question 4:**</u> Nous avons refait les questions précédentes avec des populations de tailles différentes, $5$ individus et $5000$ individus par exemple.
</pr>

In [127]:
#Pour 5 individus
liste_sequence = genere_liste_sequence(pi_0, matrice_transition_modele1, 5, 150)
liste_comptage = comptage_liste_sequence(liste_sequence)

print("La liste de 5 séquences est :\n")
print(liste_sequence)

print("\n\nLa liste de 5 séquences compte cet effectif au cours du temps:\n")
print(liste_comptage)

La liste de 5 séquences est :

[[0 0 0 0 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
  1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2]
 [0 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 

In [128]:
#affiche_graphe_nb_individu(liste_comptage)

<img src="Graph_Partie2_4.png"  align="bottom">

In [129]:
#Pour 5000 individus
liste_sequence = genere_liste_sequence(pi_0, matrice_transition_modele1, 5000, 150)
liste_comptage = comptage_liste_sequence(liste_sequence)

print("La liste de 5000 séquences est :\n")
print(liste_sequence)

print("\n\nLa liste de 5000 séquences compte cet effectif au cours du temps:\n")
print(liste_comptage)

La liste de 5000 séquences est :

[[0 0 0 ... 2 2 2]
 [0 0 1 ... 2 2 2]
 [0 0 1 ... 2 2 2]
 ...
 [0 0 0 ... 2 2 2]
 [1 1 1 ... 2 2 2]
 [0 0 0 ... 2 2 2]]


La liste de 5000 séquences compte cet effectif au cours du temps:

[[4524  476    0]
 [4130  824   46]
 [3785 1109  106]
 [3512 1305  183]
 [3254 1472  274]
 [2998 1613  389]
 [2783 1731  486]
 [2565 1831  604]
 [2361 1909  730]
 [2163 1973  864]
 [1989 2010 1001]
 [1831 2027 1142]
 [1688 2026 1286]
 [1562 2022 1416]
 [1436 2008 1556]
 [1322 1960 1718]
 [1221 1912 1867]
 [1128 1879 1993]
 [1040 1836 2124]
 [ 967 1793 2240]
 [ 895 1732 2373]
 [ 834 1659 2507]
 [ 770 1629 2601]
 [ 705 1588 2707]
 [ 645 1541 2814]
 [ 593 1486 2921]
 [ 536 1449 3015]
 [ 498 1380 3122]
 [ 459 1314 3227]
 [ 423 1247 3330]
 [ 390 1195 3415]
 [ 346 1160 3494]
 [ 311 1106 3583]
 [ 286 1038 3676]
 [ 270  985 3745]
 [ 248  939 3813]
 [ 230  890 3880]
 [ 212  856 3932]
 [ 194  822 3984]
 [ 180  775 4045]
 [ 162  749 4089]
 [ 151  689 4160]
 [ 135  653 4212]
 [ 

In [130]:
liste_pourcentage = pourcentage_liste_sequence(liste_sequence)
print("\n\nLa liste de 5000 séquences compte ces pourcentage d'individus sains/infectés/guéris au cours du temps:")
print(liste_pourcentage)



La liste de 5000 séquences compte ces pourcentage d'individus sains/infectés/guéris au cours du temps:
[[9.048e+01 9.520e+00 0.000e+00]
 [8.260e+01 1.648e+01 9.200e-01]
 [7.570e+01 2.218e+01 2.120e+00]
 [7.024e+01 2.610e+01 3.660e+00]
 [6.508e+01 2.944e+01 5.480e+00]
 [5.996e+01 3.226e+01 7.780e+00]
 [5.566e+01 3.462e+01 9.720e+00]
 [5.130e+01 3.662e+01 1.208e+01]
 [4.722e+01 3.818e+01 1.460e+01]
 [4.326e+01 3.946e+01 1.728e+01]
 [3.978e+01 4.020e+01 2.002e+01]
 [3.662e+01 4.054e+01 2.284e+01]
 [3.376e+01 4.052e+01 2.572e+01]
 [3.124e+01 4.044e+01 2.832e+01]
 [2.872e+01 4.016e+01 3.112e+01]
 [2.644e+01 3.920e+01 3.436e+01]
 [2.442e+01 3.824e+01 3.734e+01]
 [2.256e+01 3.758e+01 3.986e+01]
 [2.080e+01 3.672e+01 4.248e+01]
 [1.934e+01 3.586e+01 4.480e+01]
 [1.790e+01 3.464e+01 4.746e+01]
 [1.668e+01 3.318e+01 5.014e+01]
 [1.540e+01 3.258e+01 5.202e+01]
 [1.410e+01 3.176e+01 5.414e+01]
 [1.290e+01 3.082e+01 5.628e+01]
 [1.186e+01 2.972e+01 5.842e+01]
 [1.072e+01 2.898e+01 6.030e+01]
 [9.

In [131]:
affiche_graphe_nb_individu(liste_comptage)

<img src="Graph_Partie2_5.png"  align="bottom">

### <u>Longueur de l’infection</u>
<pr>

<u>**Question 1:**</u> À partir des simulations, nous estimons la longueur moyenne d’une séquence de $I$.
</pr>

In [132]:
def longueur_moyenne(liste_sequence, categorie):
    moy = 0
    for seq in liste_sequence:
        for etat in seq:
            if etat == categorie:
                moy += 1
    return moy/len(liste_sequence)

longueur_moyenne_infectes = longueur_moyenne(liste_sequence, 1)

print("La longueur moyenne d'une séquence de I est de", longueur_moyenne_infectes, "jours")

La longueur moyenne d'une séquence de I est de 14.2222 jours


<u>**Question 2:**</u> Affichons la distribution observée de la longueur d’infection à l'aide d'un histogramme.
</pr>

In [133]:
def repartition_longueur_moyenne(liste_sequence, categorie):
    dico_repart = {}
    for seq in liste_sequence:
        cpt = 0
        for etat in seq:
            if etat == categorie:
                cpt += 1
        if cpt in dico_repart:
            dico_repart[cpt] += 1
        else:
            dico_repart[cpt] = 1
    for i in range(max(dico_repart)):
        if i not in dico_repart:
            dico_repart[i] = 0
    
    return dico_repart

def affiche_graphe_repartition(dico_repart):
    #on crée les abscisses
    abscisses = [key for key in dico_repart]
    abscisses.sort()
    
    #on crée les ordonnées
    ordonnees = [dico_repart[key] for key in abscisses]
    
    #on plot l'histogramme
    plt.bar(abscisses, ordonnees)
    
    #on nomme l'axe des abscisses
    plt.xlabel("Longueur de l'infection (en jours)")
    #on nomme l'axe des ordonnées
    plt.ylabel("Nombre de séquences")
    
    #on paramètre l'échelle des abscisses suivant l'option x_scale
    plt.xscale("linear")
    #on paramètre l'échelle des ordonnées comme linéaire
    plt.yscale("linear")
    
    #on donne un titre au graphique
    plt.title("Histogramme de la répartition de la longueur de l'infection")
       
    #on affiche le graphe
    plt.show()

dico_repart = repartition_longueur_moyenne(liste_sequence, 1)

#affiche_graphe_repartition(dico_repart)

<img src="Histo_Partie2.png"  align="bottom">

<u>**Question 3:**</u> Calculez la longueur théorique d’une séquence de $I$ (vous pourrez utiliser l’espérance de la loi géométrique en justifiant).
</pr>

In [134]:
p = matrice_transition_modele1[1][2]
print("L'espérance de la longueur d'une séquence de I est", (1/p))
#Il faudrait ajouter la justification

L'espérance de la longueur d'une séquence de I est 14.285714285714285


<u>**Question 4:**</u> Comparer la longueur estimée et la longueur théorique.
</pr>

<pr>
La longueur estimée est de $14.23$ jours, tandis que la longueur théorique est de $14.29$ jours.<br>
On voit donc que les deux valeurs coïncident totalement.
</pr>

<u>**Question 5:**</u> Affichez la distribution théorique de la longueur d’infection.
</pr>

### <u>Petites modifications autour de ce premier modèle</u>
<pr>Vous pouvez maintenant modifier le modèle pour étudier différents cas de figure, en
faisant varier certains paramètres :
</pr>

<u>**Question 1:**</u> Faire varier la taille de la population.
</pr>

In [135]:
liste_sequence_1 = genere_liste_sequence(pi_0, matrice_transition_modele1, 100, 150)
liste_pourcentage_1 = pourcentage_liste_sequence(liste_sequence_1)

liste_sequence_2 = genere_liste_sequence(pi_0, matrice_transition_modele1, 500, 150)
liste_pourcentage_2 = pourcentage_liste_sequence(liste_sequence_2)

liste_sequence_3 = genere_liste_sequence(pi_0, matrice_transition_modele1, 1000, 150)
liste_pourcentage_3 = pourcentage_liste_sequence(liste_sequence_3)

affiche_graphe_multi(liste_pourcentage_1)
affiche_graphe_multi(liste_pourcentage_2)
affiche_graphe_multi(liste_pourcentage_3)

<pr>
Taille de la population : $100$ <img src="Graph_Partie2_6_1.png" align="bottom"><br>
Taille de la population : $500$ <img src="Graph_Partie2_6_2.png" align="bottom"><br>
Taille de la population : $1000$ <img src="Graph_Partie2_6_3.png" align="bottom">
</pr>

<pr>
On remarque donc que plus la population augmente, plus l'on tend vers les prévisions théoriques du modèle.
</pr>

<u>**Question 2:**</u> Faire varier la distribution de probabilité initiale.
</pr>

In [138]:
pi_0_2 = [0.50, 0.50, 0]
pi_0_3 = [0.10, 0.90, 0]

liste_sequence_1 = genere_liste_sequence(pi_0, matrice_transition_modele1, 1000, 150)
liste_pourcentage_1 = pourcentage_liste_sequence(liste_sequence_1)

liste_sequence_2 = genere_liste_sequence(pi_0_2, matrice_transition_modele1, 1000, 150)
liste_pourcentage_2 = pourcentage_liste_sequence(liste_sequence_2)

liste_sequence_3 = genere_liste_sequence(pi_0_3, matrice_transition_modele1, 1000, 150)
liste_pourcentage_3 = pourcentage_liste_sequence(liste_sequence_3)

affiche_graphe_multi(liste_pourcentage_1)
affiche_graphe_multi(liste_pourcentage_2)
affiche_graphe_multi(liste_pourcentage_3)

<pr>
$
π_0 = 
\begin{pmatrix}
   S &    I &    R \\
0.90 & 0.10 &    0 \\
\end{pmatrix}
$ <img src="Graph_Partie2_7_1.png" align="bottom"><br>
$
π_0 = 
\begin{pmatrix}
   S &    I &    R \\
0.50 & 0.50 &    0 \\
\end{pmatrix}
$
<img src="Graph_Partie2_7_2.png" align="bottom"><br>
$
π_0 = 
\begin{pmatrix}
   S &    I &    R \\
0.10 & 0.90 &    0 \\
\end{pmatrix}
$
<img src="Graph_Partie2_7_3.png" align="bottom">
</pr>

<pr>
On remarque que $π_0$ n'a aucune influence sur la convergence du modèle. <br>Cependant, il permet de modifier le point de départ des 3 courbes, modifiant ainsi la forme du graphique.
</pr>

<u>**Question 3:**</u> Faire varier les probabilités de transition.
</pr>

In [140]:
matrice_transition_modele1_2 = np.array([[0.97, 0.03,    0], 
                                       [   0, 0.97, 0.03], 
                                       [   0,    0,    1]])
matrice_transition_modele1_3 = np.array([[0.60, 0.40,    0], 
                                       [   0, 0.60, 0.40], 
                                       [   0,    0,    1]])

liste_sequence_1 = genere_liste_sequence(pi_0, matrice_transition_modele1, 1000, 150)
liste_pourcentage_1 = pourcentage_liste_sequence(liste_sequence_1)

liste_sequence_2 = genere_liste_sequence(pi_0, matrice_transition_modele1_2, 1000, 150)
liste_pourcentage_2 = pourcentage_liste_sequence(liste_sequence_2)

liste_sequence_3 = genere_liste_sequence(pi_0, matrice_transition_modele1_3, 1000, 150)
liste_pourcentage_3 = pourcentage_liste_sequence(liste_sequence_3)

affiche_graphe_multi(liste_pourcentage_1)
affiche_graphe_multi(liste_pourcentage_2)
affiche_graphe_multi(liste_pourcentage_3)

<pr>
$A = 
\begin{pmatrix}
  &    S &    I &    R \\
S & 0.92 & 0.08 &    0 \\
I &    0 & 0.93 & 0.07 \\
R &    0 &    0 &    1 \\
\end{pmatrix}
$<img src="Graph_Partie2_8_1.png" align="bottom"><br>
$A = 
\begin{pmatrix}
  &    S &    I &    R \\
S & 0.97 & 0.03 &    0 \\
I &    0 & 0.97 & 0.03 \\
R &    0 &    0 &    1 \\
\end{pmatrix}
$
<img src="Graph_Partie2_8_2.png" align="bottom"><br>
$A = 
\begin{pmatrix}
  &    S &    I &    R \\
S & 0.60 & 0.40 &    0 \\
I &    0 & 0.60 & 0.40 \\
R &    0 &    0 &    1 \\
\end{pmatrix}
$
<img src="Graph_Partie2_8_3.png" align="bottom">
</pr>

<pr>
On remarque que la matrice de transition (en respectant le modèle) a une grande influence sur la vitesse de convergence du modèle.<br>
En effet, lorsque les probabilités de rester sur le même état augmente, on tend plus lentement vers $π^*$.<br>
Tandis que dans le cas contraire, on tend extrêmement vite vers $π^*$.<br>
Cependant, on voit bien que modifier sur les probabilités de transition (en respectant le modèle) n'a aucune influence sur la convergence du modèle (on tend toujours vers le même état stable).
</pr>

## III) <u>Description du second modèle</u>
<pr>Nous allons maintenant considérer un second modèle, les individus guéris peuvent redevenir sains avec une probabilité de $0.02$. Ils-elles peuvent perdre leur immunité face à la maladie.
</pr>

### <u>Analyse du modèle</u>

<pr>

<u>**Question 1:**</u> Ce processus peut-il être modélisé par une chaîne de Markov ?
</pr>

<pr>
Oui (A JUSTIFIER)
</pr>

<pr>

<u>**Question 2:**</u>
Donnez la nouvelle matrice de transition.
</pr>

Matrice de transition du second modèle :<br>
$A = 
\begin{pmatrix}
  &    S &    I &    R \\
S & 0.92 & 0.08 &    0 \\
I &    0 & 0.93 & 0.07 \\
R & 0.02 &    0 & 0.98 \\
\end{pmatrix}
$

In [144]:
matrice_transition_modele2 = np.array([[0.92, 0.08,    0], 
                                       [   0, 0.93, 0.07], 
                                       [0.02,    0, 0.98]])

<pr>
    
<u>**Question 3:**</u> Quels est la nature des états de cette chaîne de Markov ? Est-elle périodique ? Est-elle irréductible ?
</pr>

<pr>
    
<u>**Question 4:**</u> Calculez la matrice $A × A$. À quoi correspond-elle ? Est elle stochastique ? Même question pour $A^3$ et $A^4$.
</pr>

<pr>
MANQUE LE "À quoi correspond-elle"<br>
$
A^2 = 
\begin{pmatrix}
  &      S &      I &      R \\
S & 0.8464 &  0.148 & 0.0056 \\
I & 0.0014 & 0.8649 & 0.1337 \\
R &  0.038 & 0.0016 & 0.9604 \\
\end{pmatrix}
$
<br><br>
$
A^3 = 
\begin{pmatrix}
  &        S &        I &        R \\
S &   0.7788 & 0.205352 & 0.191569 \\
I & 0.003962 & 0.804469 &     0.07 \\
R & 0.054168 & 0.004528 & 0.941304 \\
\end{pmatrix}
$
<br><br>
$
A^4 = 
\begin{pmatrix}
  &          S &          I &          R \\
S & 0.71681296 & 0.25328136 & 0.02990568 \\
I & 0.00747642 & 0.74847313 & 0.24405045 \\
R & 0.06866064 & 0.00854448 & 0.92279488 \\
\end{pmatrix}
$
</pr>

In [141]:
A_pow_2 = np.matmul(matrice_transition_modele2, matrice_transition_modele2)
A_pow_3 = np.matmul(A_pow_2, matrice_transition_modele2)
A_pow_4 = np.matmul(A_pow_3, matrice_transition_modele2)

print("La matrice A^2 est-elle stochastique ? :", est_stochastique(A_pow_2))
print("La matrice A^3 est-elle stochastique ? :", est_stochastique(A_pow_3))
print("La matrice A^4 est-elle stochastique ? :", est_stochastique(A_pow_4))

La matrice A^2 est-elle stochastique ? : True
La matrice A^3 est-elle stochastique ? : True
La matrice A^4 est-elle stochastique ? : True


<pr>
    
<u>**Question 5:**</u> Réalisez les nouvelles simulations. Comment la population évolue-t-elle ?
</pr>

In [146]:
liste_sequence_mod2 = genere_liste_sequence(pi_0, matrice_transition_modele2, 5000, 200)
liste_pourcentage_mod2 = pourcentage_liste_sequence(liste_sequence_mod2)

affiche_graphe_multi(liste_pourcentage_mod2)


<pr>
    
<u>**Question 6:**</u> Refaites les simulations avec une autre distribution de probabilité initiale (par exemple si au temps $t = 0$, nous avons $90\%$ d’infectés et $10\%$ de sains). Explorez d’autres initialisations et comme pour chaque question commentez vos observations.
</pr>

<pr>
    
<u>**Question 7:**</u> Calculez la distribution de probabilité stationnaire et comparez ce résultat avec les simulations pour
$t$ assez grand.
</pr>

### <u>Longueur de l’immunité</u>

<pr>
    
On peut se demander combien de temps un individu qui a été malade, reste protégé contre la maladie.
    
<u>**Question 1:**</u> À partir des simulations, estimez la longueur moyenne d’une séquence de $R$.
</pr>

<pr>
    
<u>**Question 2:**</u> Calculez la longueur théorique d’une séquence de $R$.
</pr>

<pr>
    
<u>**Question 3:**</u> Affichez la distribution théorique et la distribution observée de la longueur de l’immunité.
</pr>

### <u>Modifier le modèle</u>

<pr>
   
<u>**Question 1:**</u> Comment l’épidémie évolue-t-elle si vous modifiez la probabilité pour un individu sain de devenir infecté ? Quelle est la nouvelle distribution a l’équilibre ?
</pr>

<pr>
    
<u>**Question 2:**</u> Calculez la longueur théorique d’une séquence de $R$.
</pr>

## IV) <u>Confinement</u>
<pr>On peut imaginer que si des mesures de distanciation sociale sont mises en place, la probabilité de devenir infecté devient nulle.

Nous allons alterner entre les périodes de non distanciation et de distanciation :
- En période de non-confinement, nous utilisons la matrice de transition de l’exercice précédent
- En période de confinement, la probabilité de devenir infecté pour un individu sain devient nulle

<u>**Question 1:**</u> Commencez les simulations avec la matrice de transition de l’exercice précédent.
- On peut considérer qu’au temps initial tous les individus sont sains.
- Quand il y a $25%$ d’individus infectés dans la population, nous passons en période de distanciation. Le nombre d’individus infectés va décroître.
- Quand il y a moins de $10%$ d’infectés, le confinement est levé.
</pr>

<pr>
    
<u>**Question 2:**</u> Faites les simulations pour une population assez grande, représentez l’évolution du nombre d’individus à chaque temps $t$ (vous devriez voir des “vagues”), et notez les temps de confinement et de deconfinement.
</pr>

<pr>
    
<u>**Question 3:**</u> Combien de confinements / déconfinements sont nécessaires ?
</pr>

## V) <u>Discussion</u>

<pr>

<u>**Question 1:**</u> Quelles remarques critiques pouvez faire sur les modèles utilisés ?
</pr>

<pr>
    
<u>**Question 2:**</u> <em>(Optionnelle)</em> Proposez des améliorations et implémentez-les si possible.
</pr>